### PR TITLE
add example of creating systems via closures

### PR DESCRIPTION
**Pull request by cab**
Prompted by [this discord discussion](https://discord.com/channels/691052431525675048/691052431974465548/845026437539037274).

This example shows how to pass an arbitrary argument to a closure, and how to then use that closure as a `bevy_ecs::system::System`.

### CODE REVIEW

**Comment by mockersf** [comment]
could you also add the example to https://github.com/bevyengine/bevy/blob/main/examples/README.md?

File: examples/ecs/system_closures.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+use bevy::prelude::*;
+
+fn main() {
+    let arg = "hello";
+    App::build()
+        .add_system(Box::new(move |mut cmd: Commands| {
+            println!("this system uses the moved arg: {:?}", arg);
```
Comment (mockersf) [inline examples/ecs/system_closures.rs:7]:
```suggestion
            info!("this system uses the moved arg: {:?}", arg);
```
log rather than `println!` please 😃

**Comment by mockersf** [comment]
For the first system, you can use a `Local` parameter and config it rather than using `move` to capture `arg`:
```rust
Box::new(|mut cmd: Commands, arg: Local<String>| {
    info!("this system uses the moved arg: {:?}", arg);
    let id = cmd.spawn().id();
    info!("also it spawned an entity {:?}", id);
})
.system()
.config(|config| config.1 = Some("hello".to_string()))
```

**Comment by cab** [comment]
Thanks for the review!

> To pass parameters to your systems, you can use a `Local` parameter and config it rather than using `move` to capture `arg`:
> 
> ```rust
> Box::new(|mut cmd: Commands, arg: Local<String>| {
>     info!("this system uses the moved arg: {:?}", arg);
>     let id = cmd.spawn().id();
>     info!("also it spawned an entity {:?}", id);
> })
> .system()
> .config(|config| config.1 = Some("hello".to_string()))
> ```

Interesting, I didn't know about this. Do you think it's still useful as an example? (cc @alice-i-cecile)

**Comment by mockersf** [comment]
> Do you think this PR is still useful as an example?

Yes! We don't show how to pass closures as system, so that's valuable. And it would be even more with the `Local`/`config` way, as we don't have an example on that yet

File: examples/README.md [hunk 2]
```diff
@@ -159,6 +159,7 @@ Example | File | Description
 `startup_system` | [`ecs/startup_system.rs`](./ecs/startup_system.rs) | Demonstrates a startup system (one that runs once when the app starts up)
 `state` | [`ecs/state.rs`](./ecs/state.rs) | Illustrates how to use States to control transitioning from a Menu state to an InGame state
 `system_chaining` | [`ecs/system_chaining.rs`](./ecs/system_chaining.rs) | Chain two systems together, specifying a return type in a system (such as `Result`)
+`system_closures` | [`ecs/system_closures.rs`](./ecs/system_closures.rs) | Shows how to create systems out of closures, including how to pass arbitrary arguments via `move` or `Local`.
```
Comment (cab) [inline examples/README.md:162]: Is this too wordy?

File: Cargo.toml [hunk 3]
```diff
@@ -287,6 +287,11 @@ path = "examples/ecs/startup_system.rs"
 name = "state"
 path = "examples/ecs/state.rs"
 
+[[example]]
+name = "system_closures"
+path = "examples/ecs/system_closures.rs"
+
+
```
Comment (NathanSWard) [inline Cargo.toml:294]:
```suggestion
```

File: examples/README.md (hunk 2 above)
Comment (NathanSWard) [reply examples/README.md:162]: You could probably just reduce it to `Shows how to create and execute systems from closures` (or something to that extent)

File: examples/ecs/system_closures.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+use bevy::prelude::*;
+
+fn main() {
+    App::build()
+        .add_system(Box::new(move |mut cmd: Commands, arg: Local<String>| {
```
Comment (NathanSWard) [inline examples/ecs/system_closures.rs:5]: This is more of a question, but is the `move` necessary here?

File: examples/ecs/system_closures.rs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            info!("this system uses an argument: {:?}", arg);
+            let id = cmd.spawn().id();
+            info!("also it spawned an entity {:?}", id);
+        }).system().config(|config| config.1 = Some("hello".to_string())))
```
Comment (NathanSWard) [inline examples/ecs/system_closures.rs:9]:
```suggestion
        }).system())
```
Unless the `.config()` is completely necessary here, I feel like it's just adding unnecessary noise.

File: examples/ecs/system_closures.rs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+fn normal_system(resource: Res<ExampleResource>) {
+    info!("resource has value {:?}", resource.0);
+}
+
+pub struct ExampleResource(usize);
+
+// Creates a system that modifies a resource.
+fn create_system(arg: usize) -> impl bevy::ecs::system::System<In = (), Out = ()> {
+    Box::new(move |mut resource: ResMut<ExampleResource>| {
+        info!("system running with arg {}", arg);
```
Comment (NathanSWard) [inline examples/ecs/system_closures.rs:25]:
```suggestion
        info!("system running with captured arg {}", arg);
```

**Review by NathanSWard — CHANGES_REQUESTED** [review]

File: examples/ecs/system_closures.rs (hunk 4 above)
Comment (cab) [reply examples/ecs/system_closures.rs:5]: good catch, it's no longer necessary. will remove

File: examples/ecs/system_closures.rs (hunk 5 above)
Comment (cab) [reply examples/ecs/system_closures.rs:9]: s tI think it is necessary -- that's what sets the `Local<String>`.

**Review by NathanSWard — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
The content itself looks solid, it but could use a bit more motivating text at the start, explaining why you might want to do this.

**Comment by cab** [comment]
> The content itself looks solid, it but could use a bit more motivating text at the start, explaining why you might want to do this.

Just to make sure I'm understanding -- should I add a comment to the example, or do you mean add more explanation to the PR description?

**Comment by alice-i-cecile** [comment]
The former please; it's much more important that end users can understand the motivation. And the reviewers can check the comments in the code ;)

**Comment by cab** [comment]
sorry for dropping this! will try to wrap it up by the end of this week.

**Comment by mockersf** [comment]
as seen on Discord, `.system()` removal may not yet work with `.config()`, we should do a full pass on Bevy first

**Comment by Ratysz** [comment]
#2422 should help with that.

**Comment by alice-i-cecile** [comment]
Hey @cab I'm interested in merging this. Are you interested in refreshing this?

If so, please comment in https://github.com/bevyengine/bevy/issues/2373 and refresh this to main :)

**Comment by alice-i-cecile** [comment]
Closing due to no response on relicensing.

**Comment by alice-i-cecile** [comment]
Closing to let someone else pick this work up :) Please credit the author of this PR if you build off it.
